plugin-copilot/src/main/java/appland/copilotChat/NavieCopilotChatRequestHandler.java

```diff
@@ -0,0 +1,266 @@
+package appland.copilotChat;
+
+import appland.copilotChat.copilot.*;
+import appland.copilotChat.openAI.*;
+import com.intellij.openapi.diagnostic.Logger;
+import com.intellij.openapi.util.text.StringUtil;
+import com.intellij.util.Urls;
+import io.netty.buffer.Unpooled;
+import io.netty.channel.ChannelFutureListener;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.handler.codec.http.*;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+import org.jetbrains.ide.BuiltInServerManager;
+import org.jetbrains.ide.HttpRequestHandler;
+import org.jetbrains.io.NettyUtil;
+import org.jetbrains.io.Responses;
+
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.util.*;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * Adds a new request handler to the IDE's built-in HTTP server.
+ */
+public class NavieCopilotChatRequestHandler extends HttpRequestHandler {
+    public static final String BASE_PATH = "/vscode/copilot";
+    private static final Logger LOG = Logger.getInstance(NavieCopilotChatRequestHandler.class);
+
+    public static @NotNull String getBaseUrl() {
+        var port = BuiltInServerManager.getInstance().getPort();
+        return Urls.newUrl("http", "127.0.0.1:" + port, NavieCopilotChatRequestHandler.BASE_PATH).toExternalForm();
+    }
+
+    @Override
+    public boolean isSupported(@NotNull FullHttpRequest request) {
+        return request.method() == HttpMethod.POST;
+    }
+
+    @Override
+    public boolean process(@NotNull QueryStringDecoder queryStringDecoder,
+                           @NotNull FullHttpRequest fullHttpRequest,
+                           @NotNull ChannelHandlerContext channelHandlerContext) {
+        String requestPath;
+        try {
+            requestPath = queryStringDecoder.path();
+        } catch (IllegalArgumentException e) {
+            return false;
+        }
+
+        if (!requestPath.startsWith(BASE_PATH + "/") && !requestPath.equals(BASE_PATH)) {
+            return false;
+        }
+
+        var requestName = requestPath.substring(BASE_PATH.length());
+
+        if ("/chat/completions".equals(requestName)) {
+            // verify the authorization header, which must contain our random ide session id
+            var expectedHeaderValue = "Bearer " + GitHubCopilotService.RandomIdeSessionId;
+            if (!StringUtil.equals(fullHttpRequest.headers().get(HttpHeaderNames.AUTHORIZATION), expectedHeaderValue)) {
```

Comment:
Note while perhaps it doesn't matter terribly here, using standard string comparison to compare cryptographic information is a bad idea; because regular string comparison short circuits on first mismatch, an attacker can measure the time to rejection and learn which character was wrong. This allows bruteforcing character-by-character basically defeating any security.

It's a good idea to always use constant-time comparison when dealing with thing like this. [StackOverflow seems to suggest](https://stackoverflow.com/a/37737277) that `java.security.MessageDigest.isEqual` is what you need.